File: java_practice/MergeInterval.java
package java_practice;

import java.util.*;

public class MergeInterval {

    public static int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        List<int[]> res = new ArrayList<>();
        int[] cur = intervals[0];
        res.add(cur);

        for(int i = 1; i < intervals.length; i++){
            int[] _interval = intervals[i];
            // if overlap, merge
            if(_interval[0] <= cur[1]){
                cur[1] = Math.max(cur[1], _interval[1]);
            }else{
                // no overlap, add new interval
                cur = _interval;
                res.add(cur);
            }
        }
        return res.toArray(new int[res.size()][]);
    }

    public static void main(String[] args) {
        int[][][] tests = {
                { { 1, 3 }, { 2, 6 }, { 8, 10 }, { 15, 18 } },
                { { 1, 4 }, { 4, 5 } }
        };

        for (int[][] testcase : tests) {
            System.out.println(Arrays.deepToString(testcase) + " >>>>>>> " + Arrays.deepToString(merge(testcase)));
        }
    }
}

File: js_practice/merge_interval.js
const mergeIntervals = (intervals) => {
  intervals.sort((a, b) => a[0] - b[0]);
  const res = [];

  let cur_interval = intervals[0];

  for (let i = 1; i < intervals.length; i++) {
    const _interval = intervals[i];
    // merge into cur_interval
    if (cur_interval[1] >= _interval[0]) {
      cur_interval = [cur_interval[0], Math.max(_interval[1], cur_interval[1])];
    } else {
      // no overlap => push to result
      res.push(cur_interval);
      cur_interval = _interval;
    }
  }
  res.push(cur_interval);
  return res;
};
